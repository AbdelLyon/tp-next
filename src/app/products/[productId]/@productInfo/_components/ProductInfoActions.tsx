"use client";
import { Button } from "@/components/ui/button";
import { ShoppingCart, ArrowRight } from "lucide-react";

export const ProductInfoActions = () => {
  return (
    <div className="grid grid-cols-2 gap-4 mt-8" data-testid="product-actions">
      <Button
        size="lg"
        className="font-medium bg-primary hover:opacity-80 shadow-sm hover:shadow transition-all duration-300 rounded-lg"
        data-testid="add-to-cart-button"
        onClick={() => console.log("Add to cart button clicked")}
      >
        <ShoppingCart className="mr-2 h-4 w-4" />
        Add to Cart
      </Button>
      <Button
        size="lg"
        variant="outline"
        className="font-medium border-primary text-primary hover:opacity-80 transition-colors duration-300 rounded-lg"
        data-testid="buy-now-button"
        onClick={() => console.log("Buy now button clicked")}
      >
        Buy Now <ArrowRight className="ml-2 h-4 w-4" />
      </Button>
    </div>
  );
};
