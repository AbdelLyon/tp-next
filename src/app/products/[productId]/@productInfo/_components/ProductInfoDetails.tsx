import { Badge } from "@/components/ui/badge";

interface ProductInfoDetailsProps {
  description: string;
  tags?: string[];
}

export const ProductInfoDetails = ({
  description,
  tags,
}: ProductInfoDetailsProps) => {
  return (
    <>
      <p
        className="text-gray-600 leading-relaxed"
        data-testid="product-description"
      >
        {description}
      </p>

      {tags && tags.length > 0 && (
        <div className="flex flex-wrap gap-2 mt-5" data-testid="product-tags">
          {tags.map((tag) => (
            <Badge
              key={tag}
              variant="secondary"
              className="bg-gray-100 hover:bg-gray-200 text-gray-700 transition-colors"
            >
              {tag}
            </Badge>
          ))}
        </div>
      )}
    </>
  );
};
