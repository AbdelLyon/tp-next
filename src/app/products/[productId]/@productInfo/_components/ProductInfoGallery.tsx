import Image from "next/image";
import { Badge } from "@/components/ui/badge";
import { Button } from "x-react/button";
import { IconHeart } from "x-react/icons";

interface ProductInfoGalleryProps {
  thumbnail?: string;
  title: string;
  discountPercentage?: number;
}

export const ProductInfoGallery = ({
  thumbnail,
  title,
  discountPercentage,
}: ProductInfoGalleryProps) => {
  return (
    <div
      className="relative aspect-square rounded-xl overflow-hidden bg-background border border-border dark:bg-content1"
      data-testid="product-gallery"
    >
      {thumbnail && (
        <Image
          src={thumbnail}
          alt={title}
          fill
          className="object-cover transition-transform duration-700 group-hover:scale-105"
          sizes="(max-width: 768px) 100vw, 50vw"
          priority
          unoptimized={true}
        />
      )}

      {discountPercentage && (
        <Badge
          className="absolute top-3 left-3 bg-gradient-to-r from-red-500 to-red-600 border-0 px-3 py-1 text-sm font-medium shadow-sm"
          data-testid="discount-badge"
        >
          {discountPercentage.toFixed(0)}% OFF
        </Badge>
      )}

      <Button
        variant="light"
        className="absolute top-3 right-3 size-10 rounded-full opacity-90 hover:opacity-100 border border-border shadow-sm transition-all duration-300 hover:scale-110"
        data-testid="favorite-button"
        radius="full"
      >
        <IconHeart className="h-4 w-4 text-primary" />
      </Button>
    </div>
  );
};
