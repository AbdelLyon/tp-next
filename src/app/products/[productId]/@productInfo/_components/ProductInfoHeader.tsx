import { Badge } from "@/components/ui/badge";
import { Star, Check } from "lucide-react";
import { cn } from "@/lib/utils";

interface ProductInfoHeaderProps {
  brand?: string;
  title: string;
  rating?: number;
  stockLevel: "high" | "medium" | "low" | "out" | null;
  stock?: number;
}

export const ProductInfoHeader = ({
  brand,
  title,
  rating,
  stockLevel,
  stock,
}: ProductInfoHeaderProps) => {
  return (
    <div className="flex flex-col space-y-2 mb-4" data-testid="product-header">
      {brand && (
        <span
          className="text-sm font-medium text-primary/80"
          data-testid="product-brand"
        >
          {brand}
        </span>
      )}

      <h1
        className="text-3xl font-bold tracking-tight bg-clip-text text-transparent bg-gradient-to-r from-gray-800 to-gray-600"
        data-testid="product-title"
      >
        {title}
      </h1>

      <div className="flex items-center gap-2 mt-1">
        {rating && (
          <div className="flex items-center" data-testid="product-rating">
            {[...Array(5)].map((_, i) => {
              const difference = rating - i;
              return (
                <Star
                  key={i}
                  className={cn(
                    "h-4 w-4",
                    difference >= 1
                      ? "text-amber-400 fill-amber-400"
                      : difference >= 0.5
                      ? "text-amber-400 fill-amber-400 opacity-60"
                      : "text-gray-200",
                  )}
                />
              );
            })}
            <span className="text-sm ml-1.5 text-gray-600 font-medium">
              {rating.toFixed(1)}
            </span>
          </div>
        )}

        {stockLevel && (
          <Badge
            variant={stockLevel === "out" ? "destructive" : "outline"}
            className={cn(
              "ml-2",
              stockLevel === "high" &&
                "bg-green-50 text-green-700 border-green-200",
              stockLevel === "medium" &&
                "bg-blue-50 text-blue-700 border-blue-200",
              stockLevel === "low" &&
                "bg-amber-50 text-amber-700 border-amber-200",
            )}
            data-testid="stock-badge"
          >
            {stockLevel === "high" ? (
              <>
                <Check className="inline-block h-3 w-3 mr-1" /> In stock
              </>
            ) : stockLevel === "medium" ? (
              `In Stock (${stock})`
            ) : stockLevel === "low" ? (
              `Low Stock (${stock})`
            ) : (
              "Out of stock"
            )}
          </Badge>
        )}
      </div>
    </div>
  );
};
