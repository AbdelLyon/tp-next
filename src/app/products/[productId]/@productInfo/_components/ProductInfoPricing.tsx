import { Separator } from "@/components/ui/separator";

interface ProductInfoPricingProps {
  price: number;
  originalPrice?: number;
  discountPercentage?: number;
}

export const ProductInfoPricing = ({
  price,
  originalPrice,
  discountPercentage,
}: ProductInfoPricingProps) => {
  return (
    <>
      <div
        className="mt-4 flex items-baseline gap-3"
        data-testid="product-pricing"
      >
        <span className="text-3xl font-bold" data-testid="current-price">
          ${price.toFixed(2)}
        </span>
        {originalPrice && originalPrice > 0 && (
          <span className="text-lg line-through" data-testid="original-price">
            ${originalPrice.toFixed(2)}
          </span>
        )}
        {discountPercentage && originalPrice && (
          <span
            className="text-sm text-green-600 font-semibold bg-content2 px-2 py-0.5 rounded"
            data-testid="savings"
          >
            Save ${(originalPrice - price).toFixed(2)}
          </span>
        )}
      </div>

      <Separator className="my-5 bg-gradient-to-r from-transparent via-gray-200 to-transparent" />
    </>
  );
};
