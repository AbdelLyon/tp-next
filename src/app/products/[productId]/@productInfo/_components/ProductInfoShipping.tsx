import { Truck, Shield, Package } from "lucide-react";

export const ProductInfoShipping = () => {
  return (
    <div
      className="mt-8 space-y-3 bg-gray-50 p-4 rounded-lg border border-gray-100"
      data-testid="shipping-info"
    >
      <div className="flex items-center gap-3 text-sm">
        <div className="h-8 w-8 rounded-full bg-primary/10 flex items-center justify-center text-primary">
          <Truck className="h-4 w-4" />
        </div>
        <span>Free shipping on orders over $50</span>
      </div>
      <div className="flex items-center gap-3 text-sm">
        <div className="h-8 w-8 rounded-full bg-primary/10 flex items-center justify-center text-primary">
          <Shield className="h-4 w-4" />
        </div>
        <span>30-day money-back guarantee</span>
      </div>
      <div className="flex items-center gap-3 text-sm">
        <div className="h-8 w-8 rounded-full bg-primary/10 flex items-center justify-center text-primary">
          <Package className="h-4 w-4" />
        </div>
        <span>Secure packaging and delivery</span>
      </div>
    </div>
  );
};
