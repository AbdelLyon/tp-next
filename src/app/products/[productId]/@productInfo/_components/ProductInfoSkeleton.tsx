import { Skeleton } from "@/components/ui/skeleton";
import { Separator } from "@/components/ui/separator";
import { Tabs, TabsContent, TabsList, TabsTrigger } from "@/components/ui/tabs";

export const ProductInfoSkeleton = () => {
  return (
    <div className="rounded-lg border border-default-200 shadow-sm overflow-hidden">
      <div className="grid grid-cols-1 md:grid-cols-2 gap-8 p-6">
        <div className="relative aspect-square rounded-lg overflow-hidden bg-gray-50 dark:bg-content2">
          <Skeleton className="h-full w-full" />
          <div className="absolute top-3 left-3">
            <Skeleton className="h-6 w-16 rounded-full" />
          </div>
          <div className="absolute top-3 right-3">
            <Skeleton className="h-9 w-9 rounded-full" />
          </div>
        </div>

        <div className="flex flex-col">
          <Skeleton className="h-9 w-4/5 mb-2" />
          <Skeleton className="h-5 w-1/3 mb-4" />

          <div className="flex items-center gap-2 mt-3">
            <Skeleton className="h-4 w-28" />
            <Skeleton className="h-6 w-24 rounded-full ml-2" />
          </div>

          <div className="mt-6 flex items-baseline gap-2">
            <Skeleton className="h-8 w-24" />
            <Skeleton className="h-6 w-20" />
          </div>

          <div className="mt-6 space-y-2">
            <Skeleton className="h-4 w-full" />
            <Skeleton className="h-4 w-full" />
            <Skeleton className="h-4 w-3/4" />
          </div>

          <div className="flex flex-wrap gap-2 mt-6">
            <Skeleton className="h-6 w-16 rounded-full" />
            <Skeleton className="h-6 w-20 rounded-full" />
            <Skeleton className="h-6 w-14 rounded-full" />
          </div>

          <div className="grid grid-cols-2 gap-3 mt-8">
            <Skeleton className="h-11 w-full rounded-md" />
            <Skeleton className="h-11 w-full rounded-md" />
          </div>

          <div className="mt-8 space-y-3">
            <div className="flex items-center gap-2">
              <Skeleton className="h-4 w-4" />
              <Skeleton className="h-4 w-48" />
            </div>
            <div className="flex items-center gap-2">
              <Skeleton className="h-4 w-4" />
              <Skeleton className="h-4 w-56" />
            </div>
          </div>
        </div>
      </div>

      <Separator />

      <div className="p-6">
        <Tabs defaultValue="details">
          <TabsList className="grid w-full grid-cols-3">
            <TabsTrigger value="details">Details</TabsTrigger>
            <TabsTrigger value="specifications">Specifications</TabsTrigger>
            <TabsTrigger value="reviews">Reviews</TabsTrigger>
          </TabsList>
          <TabsContent value="details" className="pt-4">
            <Skeleton className="h-6 w-40 mb-3" />
            <div className="space-y-2">
              <Skeleton className="h-4 w-full" />
              <Skeleton className="h-4 w-full" />
              <Skeleton className="h-4 w-3/4" />
            </div>
          </TabsContent>
        </Tabs>
      </div>
    </div>
  );
};
