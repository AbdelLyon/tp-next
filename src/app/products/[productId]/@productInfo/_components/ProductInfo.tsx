import { productService } from "@/services/ProductService";
import { ProductInfoGallery } from "./ProductInfoGallery";
import { ProductInfoHeader } from "./ProductInfoHeader";
import { ProductInfoPricing } from "./ProductInfoPricing";
import { ProductInfoDetails } from "./ProductInfoDetails";
import { ProductInfoActions } from "./ProductInfoActions";
import { ProductInfoShipping } from "./ProductInfoShipping";
import { ProductInfoTabs } from "./ProductInfoTabs";

const ProductInfo = async ({ productId }: { productId: string }) => {
  const product = await productService.getProductById(productId);

  const originalPrice = product.discountPercentage
    ? product.price * (1 + product.discountPercentage / 100)
    : 0;

  const stockLevel =
    product.stock !== undefined
      ? product.stock > 20
        ? "high"
        : product.stock > 5
        ? "medium"
        : product.stock > 0
        ? "low"
        : "out"
      : null;

  return (
    <>
      <div className="grid grid-cols-1 md:grid-cols-2 gap-8 p-6">
        <ProductInfoGallery
          thumbnail={product.thumbnail}
          title={product.title}
          discountPercentage={product.discountPercentage}
        />

        <div className="flex flex-col">
          <ProductInfoHeader
            brand={product.brand}
            title={product.title}
            rating={product.rating}
            stockLevel={stockLevel}
            stock={product.stock}
          />

          <ProductInfoPricing
            price={product.price}
            originalPrice={originalPrice}
            discountPercentage={product.discountPercentage}
          />

          <ProductInfoDetails
            description={product.description}
            tags={product.tags}
          />

          <ProductInfoActions />

          <ProductInfoShipping />
        </div>
      </div>

      <div className="px-6 pb-6">
        <ProductInfoTabs product={product} />
      </div>
    </>
  );
};

export default ProductInfo;
