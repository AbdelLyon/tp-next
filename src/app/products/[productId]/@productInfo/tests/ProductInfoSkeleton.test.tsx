import { describe, it, expect, vi } from "vitest";
import { render } from "@testing-library/react";
import { ProductInfoSkeleton } from "../_components/ProductInfoSkeleton";

interface SkeletonProps {
  className?: string;
}

interface SeparatorProps {
  className?: string;
}

interface TabsProps {
  defaultValue: string;
  children: React.ReactNode;
}

interface TabsListProps {
  className?: string;
  children: React.ReactNode;
}

interface TabsTriggerProps {
  value: string;
  children: React.ReactNode;
}

interface TabsContentProps {
  value: string;
  className?: string;
  children: React.ReactNode;
}

// Mocks pour les composants utilisÃ©s
vi.mock("@/components/ui/skeleton", () => ({
  Skeleton: ({ className }: SkeletonProps) => (
    <div data-testid="skeleton" className={className}></div>
  ),
}));

vi.mock("@/components/ui/separator", () => ({
  Separator: ({ className }: SeparatorProps) => (
    <hr data-testid="separator" className={className} />
  ),
}));

vi.mock("@/components/ui/tabs", () => ({
  Tabs: ({ defaultValue, children }: TabsProps) => (
    <div data-testid="tabs" data-default-value={defaultValue}>
      {children}
    </div>
  ),
  TabsList: ({ className, children }: TabsListProps) => (
    <div data-testid="tabs-list" className={className}>
      {children}
    </div>
  ),
  TabsTrigger: ({ value, children }: TabsTriggerProps) => (
    <button data-testid={`tabs-trigger-${value}`} data-value={value}>
      {children}
    </button>
  ),
  TabsContent: ({ value, className, children }: TabsContentProps) => (
    <div
      data-testid={`tabs-content-${value}`}
      data-value={value}
      className={className}
    >
      {children}
    </div>
  ),
}));

describe("ProductInfoSkeleton", () => {
  it("devrait correspondre au snapshot", () => {
    const { container } = render(<ProductInfoSkeleton />);
    expect(container).toMatchSnapshot();
  });
});
