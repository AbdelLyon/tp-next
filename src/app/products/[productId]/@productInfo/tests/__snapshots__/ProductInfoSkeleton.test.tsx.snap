// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductInfoSkeleton > devrait correspondre au snapshot 1`] = `
<div>
  <div
    class="bg-white rounded-lg border shadow-sm overflow-hidden"
  >
    <div
      class="grid grid-cols-1 md:grid-cols-2 gap-8 p-6"
    >
      <div
        class="relative aspect-square rounded-lg overflow-hidden bg-gray-100"
      >
        <div
          class="h-full w-full"
          data-testid="skeleton"
        />
        <div
          class="absolute top-3 left-3"
        >
          <div
            class="h-6 w-16 rounded-full"
            data-testid="skeleton"
          />
        </div>
        <div
          class="absolute top-3 right-3"
        >
          <div
            class="h-9 w-9 rounded-full"
            data-testid="skeleton"
          />
        </div>
      </div>
      <div
        class="flex flex-col"
      >
        <div
          class="h-9 w-4/5 mb-2"
          data-testid="skeleton"
        />
        <div
          class="h-5 w-1/3 mb-4"
          data-testid="skeleton"
        />
        <div
          class="flex items-center gap-2 mt-3"
        >
          <div
            class="h-4 w-28"
            data-testid="skeleton"
          />
          <div
            class="h-6 w-24 rounded-full ml-2"
            data-testid="skeleton"
          />
        </div>
        <div
          class="mt-6 flex items-baseline gap-2"
        >
          <div
            class="h-8 w-24"
            data-testid="skeleton"
          />
          <div
            class="h-6 w-20"
            data-testid="skeleton"
          />
        </div>
        <div
          class="mt-6 space-y-2"
        >
          <div
            class="h-4 w-full"
            data-testid="skeleton"
          />
          <div
            class="h-4 w-full"
            data-testid="skeleton"
          />
          <div
            class="h-4 w-3/4"
            data-testid="skeleton"
          />
        </div>
        <div
          class="flex flex-wrap gap-2 mt-6"
        >
          <div
            class="h-6 w-16 rounded-full"
            data-testid="skeleton"
          />
          <div
            class="h-6 w-20 rounded-full"
            data-testid="skeleton"
          />
          <div
            class="h-6 w-14 rounded-full"
            data-testid="skeleton"
          />
        </div>
        <div
          class="grid grid-cols-2 gap-3 mt-8"
        >
          <div
            class="h-11 w-full rounded-md"
            data-testid="skeleton"
          />
          <div
            class="h-11 w-full rounded-md"
            data-testid="skeleton"
          />
        </div>
        <div
          class="mt-8 space-y-3"
        >
          <div
            class="flex items-center gap-2"
          >
            <div
              class="h-4 w-4"
              data-testid="skeleton"
            />
            <div
              class="h-4 w-48"
              data-testid="skeleton"
            />
          </div>
          <div
            class="flex items-center gap-2"
          >
            <div
              class="h-4 w-4"
              data-testid="skeleton"
            />
            <div
              class="h-4 w-56"
              data-testid="skeleton"
            />
          </div>
        </div>
      </div>
    </div>
    <hr
      data-testid="separator"
    />
    <div
      class="p-6"
    >
      <div
        data-default-value="details"
        data-testid="tabs"
      >
        <div
          class="grid w-full grid-cols-3"
          data-testid="tabs-list"
        >
          <button
            data-testid="tabs-trigger-details"
            data-value="details"
          >
            Details
          </button>
          <button
            data-testid="tabs-trigger-specifications"
            data-value="specifications"
          >
            Specifications
          </button>
          <button
            data-testid="tabs-trigger-reviews"
            data-value="reviews"
          >
            Reviews
          </button>
        </div>
        <div
          class="pt-4"
          data-testid="tabs-content-details"
          data-value="details"
        >
          <div
            class="h-6 w-40 mb-3"
            data-testid="skeleton"
          />
          <div
            class="space-y-2"
          >
            <div
              class="h-4 w-full"
              data-testid="skeleton"
            />
            <div
              class="h-4 w-full"
              data-testid="skeleton"
            />
            <div
              class="h-4 w-3/4"
              data-testid="skeleton"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
